persistence/provider/sql/sqlite/eventstore.go

```diff
@@ -289,6 +290,83 @@ func (driver) SelectEvents(
 	)
 }
 
+// SelectEventsBySource selects events from the eventstore that match the
+// given source, namely the source's key and id.
+func (driver) SelectEventsBySource(
+	ctx context.Context,
+	db *sql.DB,
+	ak, hk, id string,
+) (*sql.Rows, error) {
+	return db.QueryContext(
+		ctx,
+		`SELECT
+			e.offset,
+			e.message_id,
+			e.causation_id,
+			e.correlation_id,
+			e.source_app_name,
+			e.source_app_key,
+			e.source_handler_name,
+			e.source_handler_key,
+			e.source_instance_id,
+			e.created_at,
+			e.description,
+			e.portable_name,
+			e.media_type,
+			e.data
+		FROM event AS e
+		WHERE e.source_app_key = $1
+		AND e.source_handler_key = $2
+		AND e.source_instance_id = $3
+		ORDER BY e.offset`,
+		ak,
+		hk,
+		id,
+	)
+}
+
+// SelectEventsBySourceAfterMessage selects events from the eventstore that
+// match the given source, namely the source's key and id. The events must
+// after the specified message d.
+func (driver) SelectEventsBySourceAfterMessage(
+	ctx context.Context,
+	db *sql.DB,
+	ak, hk, id, d string,
+) (*sql.Rows, error) {
+	return db.QueryContext(
+		ctx,
+		`SELECT
+			e.offset,
+			e.message_id,
+			e.causation_id,
+			e.correlation_id,
+			e.source_app_name,
+			e.source_app_key,
+			e.source_handler_name,
+			e.source_handler_key,
+			e.source_instance_id,
+			e.created_at,
+			e.description,
+			e.portable_name,
+			e.media_type,
+			e.data
+		FROM event AS e
+		WHERE e.source_app_key = $1
+		AND e.source_handler_key = $2
+		AND e.source_instance_id = $3
+		AND e.offset > (
+			SELECT e1.offset
+			FROM event AS e1
+			WHERE e1.message_id = $4
```

Comment:
Looking at how this is implemented, I can't help but think we'd be better off having a driver method like `SelectOffsetByMessageID()`, calling that first to get the offset, then passing that offset to `SelectEventsBySource()`.

The downside is that there's an extra round trip, but it in practice it would only happen when the aggregate isn't cached or snapshotted.

The upside is it that it lets us reuse `SelectEventsBySource()` which is really the same query. More importantly, it gives us an opportunity to properly handle the failure condition that would occur if we can't actually find that message ID. It may also reduce the number of size of indices we need on the event table.

Reply:
Yeah those subqueries felt like something out-of-line when I wrote them. The only goal I was pursuing at that time was to build autonomous methods enough to break away from the`eventstore.Query` usage and that dynamic query builder that is present in every implementation of `sql.Driver.QueryEvents`. `SelectOffsetByMessageID()` still does not break that tendency and that's so awesome.

Following the proposal, `SelectEventsBySource()` would now have to have an offset parameter. That might present a problem as this parameter would essentially mean all events with the offset greater than the given. When not used ( offset param == 0) this parameter will exclude the event with zero offset unless we code it as a dynamic query similar to  `sql.Driver.QueryEvents` implementations which we are trying to avoid.

I do admit that my concerns might be flawed 😄.

Reply:
> When not used ( offset param == 0) this parameter will exclude the event with zero offset unless we code it as a dynamic query similar to `sql.Driver.QueryEvents` implementations which we are trying to avoid.

All we need to do is make the `WHERE` clause use `offset >= ?` and then add `1` to the offset of the "barrier" message, I'm pretty sure.

By the way, are you happy with that terminology - "barrier"?  I've seen it used to refer to offsets in the open/exclusive ranges we talked about with the `[x, y)` notation.

Reply:
Yeah man barrier seems logical to refer to offset limit

Reply:
It also sounds like we should return an error if the 'barier' message with the given ID is not found. I am also gonna add a test to address this case and change `boltdb` and `memory` to behave consistently.

Also, Can discuss the names of the SQL indices and their constituent fields when you have some free time?

Reply:
> Can discuss the names of the SQL indices and their constituent fields when you have some free time?

Yep, just slack call whenever.